File: lib-esphome-client/src/test/java/io/github/arhor/esphome/client/async/noise/NoiseHandshakeStateTest.java
package io.github.arhor.esphome.client.async.noise;

import io.github.arhor.esphome.client.async.internal.exception.EspHomeProtocolException;
import io.github.arhor.esphome.client.async.internal.noise.NoiseConstants;
import io.github.arhor.esphome.client.async.internal.noise.NoiseHandshakeState;
import org.junit.jupiter.api.Test;

import java.nio.charset.StandardCharsets;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

class NoiseHandshakeStateTest {

    @Test
    void initiatorAndResponderDeriveCompatibleCipherStates() {
        final var psk = new byte[32];
        for (var index = 0; index < 32; index++) {
            psk[index] = (byte) (index + 1);
        }
        final var initiator = NoiseHandshakeState.initiator(psk);
        final var responder = NoiseHandshakeState.responder(psk);

        final var firstMessage = initiator.writeMessage();
        responder.readMessage(firstMessage);
        final var secondMessage = responder.writeMessage();
        initiator.readMessage(secondMessage);

        assertTrue(initiator.isComplete());
        assertTrue(responder.isComplete());

        final var clientPlaintext = "hello from client".getBytes(StandardCharsets.UTF_8);
        final var clientCiphertext = initiator.getSendCipher().encryptWithAd(NoiseConstants.EMPTY, clientPlaintext);
        assertArrayEquals(
            clientPlaintext,
            responder.getReceiveCipher().decryptWithAd(NoiseConstants.EMPTY, clientCiphertext)
        );

        final var serverPlaintext = "hello from server".getBytes(StandardCharsets.UTF_8);
        final var serverCiphertext = responder.getSendCipher().encryptWithAd(NoiseConstants.EMPTY, serverPlaintext);
        assertArrayEquals(
            serverPlaintext,
            initiator.getReceiveCipher().decryptWithAd(NoiseConstants.EMPTY, serverCiphertext)
        );
    }

    @Test
    void handshakeRejectsMismatchedPreSharedKeys() {
        final var initiatorPsk = new byte[32];
        final var responderPsk = new byte[32];
        for (var index = 0; index < 32; index++) {
            initiatorPsk[index] = (byte) (index + 1);
            responderPsk[index] = (byte) (index + 2);
        }
        final var initiator = NoiseHandshakeState.initiator(initiatorPsk);
        final var responder = NoiseHandshakeState.responder(responderPsk);

        final var firstMessage = initiator.writeMessage();

        assertThrows(EspHomeProtocolException.class, () -> responder.readMessage(firstMessage));
    }
}

File: services/lib-esphome-client/src/test/kotlin/io/github/arhor/esphome/client/internal/noise/NoiseHandshakeStateTest.kt
package io.github.arhor.esphome.client.internal.noise

import io.github.arhor.esphome.client.EspHomeProtocolException
import kotlin.test.Test
import kotlin.test.assertContentEquals
import kotlin.test.assertFailsWith
import kotlin.test.assertTrue

class NoiseHandshakeStateTest {

    @Test
    fun `initiator and responder derive compatible cipher states`() {
        val psk = ByteArray(32) { index -> (index + 1).toByte() }
        val initiator = NoiseHandshakeState.initiator(psk)
        val responder = NoiseHandshakeState.responder(psk)

        val firstMessage = initiator.writeMessage()
        responder.readMessage(firstMessage)
        val secondMessage = responder.writeMessage()
        initiator.readMessage(secondMessage)

        assertTrue(initiator.isComplete)
        assertTrue(responder.isComplete)

        val clientPlaintext = "hello from client".encodeToByteArray()
        val clientCiphertext = initiator.sendCipher.encryptWithAd(ByteArray(0), clientPlaintext)
        assertContentEquals(clientPlaintext, responder.receiveCipher.decryptWithAd(ByteArray(0), clientCiphertext))

        val serverPlaintext = "hello from server".encodeToByteArray()
        val serverCiphertext = responder.sendCipher.encryptWithAd(ByteArray(0), serverPlaintext)
        assertContentEquals(serverPlaintext, initiator.receiveCipher.decryptWithAd(ByteArray(0), serverCiphertext))
    }

    @Test
    fun `handshake rejects mismatched pre shared keys`() {
        val initiator = NoiseHandshakeState.initiator(ByteArray(32) { index -> (index + 1).toByte() })
        val responder = NoiseHandshakeState.responder(ByteArray(32) { index -> (index + 2).toByte() })

        val firstMessage = initiator.writeMessage()

        assertFailsWith<EspHomeProtocolException> {
            responder.readMessage(firstMessage)
        }
    }
}
